console/v1/types_console_plugin.go

```diff
@@ -0,0 +1,173 @@
+package v1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +openshift:compatibility-gen:level=1
+
+// ConsolePlugin is an extension for customizing OpenShift web console by
+// dynamically loading code from another service running on the cluster.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+type ConsolePlugin struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+
+	// +kubebuilder:validation:Required
+	Spec ConsolePluginSpec `json:"spec"`
+}
+
+// ConsolePluginSpec is the desired plugin configuration.
+type ConsolePluginSpec struct {
+	// displayName is the display name of the plugin.
+	// The dispalyName should be between 1 and 128 characters.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=128
+	DisplayName string `json:"displayName"`
+	// service is a Kubernetes Service that exposes the plugin using a
+	// deployment with an HTTP server. The Service must use HTTPS and
+	// Service serving certificate. The console backend will proxy the
+	// plugins assets from the Service using the service CA bundle.
+	// +kubebuilder:validation:Required
+	Service ConsolePluginService `json:"service"`
+	// proxy is a list of proxies that describe various service type
+	// to which the plugin needs to connect to.
+	// +kubebuilder:validation:Optional
+	Proxy []ConsolePluginProxy `json:"proxy,omitempty"`
+	// i18n is the configuration of plugin's localization resources.
+	// +kubebuilder:validation:Required
+	I18n ConsolePluginI18n `json:"i18n"`
+}
+
+// ConsolePluginI18n holds information on localization resources that are served by
+// the dynamic plugin.
+type ConsolePluginI18n struct {
+	// enabled indicates if the plugin contains localization resource. If enabled,
+	// i18n namespace, that the plugin contains, will be loaded by the console server.
+	// If i18n is not enabled, no i18n namespaces will be loaded for the plugin.
+	// By default loading of i18n namespace is not enabled.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:default=false
+	Enabled bool `json:"enabled"`
```

Comment:
why is this still a bool?

i think it's worth having a conversation about keeping the Authorize field as a bool in order to avoid a conversion webhook, but for a net new field, we should be following the guidance to not use bools.

Reply:
If this is new, it should be an enum as Ben has said. Let's have a conversation about conversion webhooks. There are other elements we would change than just the bool if we wanted to follow current conventions (namely theres a discriminated union in the API but it's mixed with other fields).

If everyone is available, perhaps the control plane arch call next week is a good place to discuss?